plutus-benchmark/ed25519-throughput/Main.hs

```diff
@@ -0,0 +1,213 @@
+{-# LANGUAGE NoImplicitPrelude   #-}
+{-# LANGUAGE NumericUnderscores  #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+{- | Check how many Ed25519 signature verifications we can perform within the
+   limits specified in the protocol parameters.
+-}
+
+module Main (main)
+
+where
+
+import PlutusCore (DefaultFun, DefaultUni)
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget (exBudgetCPU, exBudgetMemory))
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults qualified as PLC
+import PlutusCore.Evaluation.Machine.ExMemory (ExCPU (..), ExMemory (..))
+import PlutusTx qualified as Tx
+import UntypedPlutusCore qualified as UPLC
+import UntypedPlutusCore.Evaluation.Machine.Cek qualified as Cek
+
+import PlutusTx.IsData (toData, unstableMakeIsData)
+import PlutusTx.Prelude as Tx hiding (sort, (*))
+
+import Cardano.Crypto.DSIGN.Class (ContextDSIGN, DSIGNAlgorithm, Signable, deriveVerKeyDSIGN, genKeyDSIGN,
+                                   rawSerialiseSigDSIGN, rawSerialiseVerKeyDSIGN, signDSIGN)
+import Cardano.Crypto.DSIGN.Ed25519 (Ed25519DSIGN)
+import Cardano.Crypto.Seed (mkSeedFromBytes)
+
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.ByteString.Hash qualified as Hash
+import Flat qualified
+import Hedgehog qualified as H (Seed (..))
+import Hedgehog.Internal.Gen qualified as G
+import Hedgehog.Internal.Range qualified as R
+import Hedgehog.Internal.Tree qualified as T
+import System.IO.Unsafe (unsafePerformIO)
+import Text.Printf (printf)
+
+import Prelude (Double, IO, Integral, String, error, fromIntegral, maybe, (*), (/))
+
+type UTerm   = UPLC.Term    UPLC.NamedDeBruijn DefaultUni DefaultFun ()
+type UProg   = UPLC.Program UPLC.NamedDeBruijn DefaultUni DefaultFun ()
+type UDBProg = UPLC.Program UPLC.DeBruijn      DefaultUni DefaultFun ()
+
+
+-------------------------------- PLC stuff--------------------------------
+
+compiledCodeToTerm
+    :: Tx.CompiledCodeIn DefaultUni DefaultFun a -> UTerm
+compiledCodeToTerm (Tx.getPlc -> UPLC.Program _ _ body) = body
+
+{- | Remove the textual names from a NamedDeBruijn program -}
+toAnonDeBruijnProg :: UProg -> UDBProg
+toAnonDeBruijnProg (UPLC.Program () ver body) =
+    UPLC.Program () ver $ UPLC.termMapNames (\(UPLC.NamedDeBruijn _ ix) -> UPLC.DeBruijn ix) body
+
+
+---------------- Inputs ----------------
+
+{- | Generate n public keys P_1,...,P_n (we'll call these the "data keys") and
+   hash them to get n hashes H_1,...,H_n.  Sign all of the hashes with different
+   private keys K_1,...,K_n (with corresponding public keys V_1,...V_n) to get n
+   signatures S_1,...,S_n.  We create a list of (V_i, H_i, S_i, P_i) tuples,
+   convert it to Data, and feed that to a script which
+
+     1. Verifies each (V_i, H_i, S_i) to check that the signatures are valid.
+     2. Hashes each P_i to make sure that it matches H_i.
```

Comment:
Is this something we actually need to do? do we not just work with the hashes normally? question for @iquerejeta probably